csharp_dotnetcore/12.NLP-With-LUIS/LuisBot/AppInsights/MyAppInsightLuisRecognizer.cs

```diff
@@ -0,0 +1,124 @@
+            : base(application, predictionOptions, includeApiResults)
+        {
+            LogOriginalMessage = logOriginalMessage;
+            LogUsername = logUserName;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether to log the <see cref="Activity"/> message text that came from the user.
+        /// </summary>
+        /// <value>If true, will log the Activity Message text into the AppInsight Custome Event for Luis intents.</value>
+        public bool LogOriginalMessage { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether to log the User name.
+        /// </summary>
+        /// <value>If true, will log the user name into the AppInsight Custom Event for Luis intents.</value>
+        public bool LogUsername { get; }
+
+        /// <summary>
+        /// Analyze the current message text and return results of the analysis (Suggested actions and intents).
+        /// </summary>
+        /// <param name="context">Context object containing information for a single turn of conversation with a user.</param>
+        /// <param name="ct">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <param name="logOriginalMessage">Determines if the original message is logged into Application Insights.  This is a privacy consideration.</param>
+        /// <returns>The LUIS results of the analysis of the current message text in the current turn's context activity.</returns>
+        public async Task<RecognizerResult> RecognizeAsync(ITurnContext context, CancellationToken ct, bool logOriginalMessage = false)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Call Luis Recognizer
+            var recognizerResult = await base.RecognizeAsync(context, ct);
+
+            var conversationId = context.Activity.Conversation.Id;
+
+            // Find the Telemetry Client
+            if (context.TurnState.TryGetValue(MyAppInsightsLoggerMiddleware.AppInsightsServiceKey, out var telemetryClient) && recognizerResult != null)
+            {
+                var topLuisIntent = recognizerResult.GetTopScoringIntent();
+                var intentScore = topLuisIntent.score.ToString("N2");
+
+                // Add the intent score and conversation id properties
+                var telemetryProperties = new Dictionary<string, string>()
+                {
+                    { MyLuisConstants.ActivityIdProperty, context.Activity.Id },
+                    { MyLuisConstants.IntentProperty, topLuisIntent.intent },
+                    { MyLuisConstants.IntentScoreProperty, intentScore },
+                };
+
+                if (recognizerResult.Properties.TryGetValue("sentiment", out var sentiment) && sentiment is JObject)
+                {
+                    if (((JObject)sentiment).TryGetValue("label", out var label))
```

Comment:
Consider casting (JObject)sentiment only a single time, and doing it into a variable. It's happening twice in the code below, and would be more clear that way. It would also be easier to debug, as the hover-overs would work correctly during debugging.

Reply:
Fixed